### PR TITLE
User Delegation and Directory SAS

### PR DESCRIPTION
**Pull request by amnguye**

### CODE REVIEW

File: sdk/storage/Azure.Storage.Common/src/Sas/SasQueryParameters.cs [hunk 1]
```diff
@@ -169,6 +182,46 @@ public partial class SasQueryParameters
...
+        /// The AAD Object ID of a user authorized by the owner of the User Delegation Key
+        /// to perform the action granted by the SAS. The Azure Storage service will
+        /// ensure that the owner of the user delegation key has the required permissions
+        /// before granting access but no additional permission check for the user specified
+        /// in this value will be performed. This cannot be used in conjuction with
+        /// <see cref="UnauthorizedAadObjectId"/>.
+        /// Only valid in an HNS enabled account. If this value is set in an non-HNS enabled
+        /// account, an authorization failure will be thrown.
+        /// </summary>
+        public string AuthorizedAadObjectId => _authorizedAadObjectId ?? string.Empty;
```
Comment (gapra-msft) [inline sdk/storage/Azure.Storage.Common/src/Sas/SasQueryParameters.cs:197]: Do we want these publically exposed as authorized and unauthorized?

File: sdk/storage/Azure.Storage.Files.DataLake/src/Sas/DataLakeSasBuilder.cs [hunk 2]
```diff
@@ -134,6 +148,45 @@ public class DataLakeSasBuilder
...
+        /// The Azure Storage service will ensure that the owner of the
+        /// user delegation key has the required permissions before granting access.
+        /// If <see cref="_performPosixCheck"/> is enabled, the Azure Storage Service
+        /// will perform an additional POSIX ACL check to determine if the user is
+        /// authorized to perform the requested operation. If not enabled no additional
+        /// permission check for the user specified in this value will be performed.
+        /// This cannot be used in conjuction with. This is only used with generating
+        /// User Delegation SAS.
+        /// </summary>
+        private string _agentObjectId { get; set; }
```
Comment (gapra-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/src/Sas/DataLakeSasBuilder.cs:164]: Do we want this settable here and in the setter below?

File: sdk/storage/Azure.Storage.Files.DataLake/src/Sas/DataLakeSasBuilder.cs [hunk 3]
```diff
@@ -199,18 +252,48 @@ public void SetPermissions(string rawPermissions)
...
+        /// user delegation key has the required permissions before granting access.
+        /// If the performPosicCheck is enabled, the Azure Storage Service
+        /// will perform an additional POSIX ACL check to determine if the user is
+        /// authorized to perform the requested operation. If not enabled no additional
+        /// permission check for the user specified in this value will be performed.
+        /// This cannot be used in conjuction with. This is only used with generating
+        /// User Delegation SAS.
+        /// </summary>
+        /// <param name="objectId">
+        /// This value witll be used for the AAD Object of a user authorized by the
```
Comment (gapra-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/src/Sas/DataLakeSasBuilder.cs:269]: typo: will

File: sdk/storage/Azure.Storage.Files.DataLake/src/Sas/DataLakeSasBuilder.cs [hunk 4]
```diff
@@ -90,6 +90,15 @@ public class DataLakeSasBuilder
         /// </summary>
         public string Path { get; set; }
 
+        /// <summary>
+        /// Defines whether or not the path is a directory. If set to true,
+        /// the path will be assumed to be a directory for a directory SAS.
+        /// If set to false or default, the path will be assumed as a path
```
Comment (gapra-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/src/Sas/DataLakeSasBuilder.cs:96]: path will be assumed as a file?

Comment (gapra-msft) [reply sdk/storage/Azure.Storage.Files.DataLake/src/Sas/DataLakeSasBuilder.cs:96]: Can we add docs near the path setter about setting this as well?

File: sdk/storage/Azure.Storage.Common/src/Sas/SasQueryParameters.cs [hunk 5]
```diff
@@ -60,6 +61,18 @@ public partial class SasQueryParameters
...
+        private string _authorizedAadObjectId;
+
+        /// suoid
+        private string _unauthorizedAadObjectId;
+
+        /// scid
+        private string _correlationId;
+
+        // sdd
+        private uint? _directoryDepth;
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/src/Sas/SasQueryParameters.cs:74]: why `uint`?

File: sdk/storage/Azure.Storage.Blobs/src/Sas/BlobSasBuilder.cs [hunk 6]
```diff
@@ -349,6 +349,9 @@ public BlobSasQueryParameters ToSasQueryParameters(UserDelegationKey userDelegat
                 signedExpiry,
                 userDelegationKey.SignedService,
                 userDelegationKey.SignedVersion,
+                null, // authorized aad object id
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/src/Sas/BlobSasBuilder.cs:352]: Should we add authorized aad object id, unauthorized add object id, and correlation id as settable properties on the `BlobSasBuilder`?

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/src/Sas/BlobSasBuilder.cs:352]: These properties might be Data Lake only, in which case this comment doesn't make sense.

File: sdk/storage/Azure.Storage.Common/src/Sas/SasQueryParameters.cs (hunk 5 above)
Comment (gapra-msft) [reply sdk/storage/Azure.Storage.Common/src/Sas/SasQueryParameters.cs:74]: I think cause directory depth can only be 0 and above

File: sdk/storage/Azure.Storage.Files.DataLake/src/Sas/DataLakeSasBuilder.cs (hunk 2 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.Files.DataLake/src/Sas/DataLakeSasBuilder.cs:164]: I set the variable as private, so it should only be accessible only in the class.

File: sdk/storage/Azure.Storage.Common/src/Sas/SasQueryParameters.cs (hunk 1 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.Common/src/Sas/SasQueryParameters.cs:197]: Maybe I will wait until we can come up with better names before I go and rename this.

File: sdk/storage/Azure.Storage.Files.DataLake/src/Sas/DataLakeSasBuilder.cs (hunk 4 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.Files.DataLake/src/Sas/DataLakeSasBuilder.cs:96]: Good catch! haha "the path will be assumed as a path"

**Review by seanmcc-msft — APPROVED** [review]

**Review by kasobol-msft — APPROVED** [review]

**Pull request merged**